pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1306,3 +1312,106 @@ func (vs *VSphere) NodeManager() (nodeManager *NodeManager) {
 	}
 	return vs.nodeManager
 }
+
+// GetZone implements Zones.GetZone
+func (vs *VSphere) GetZone(ctx context.Context) (cloudprovider.Zone, error) {
+	nodeName, err := vs.CurrentNodeName(ctx, vs.hostName)
+	if err != nil {
+		glog.Errorf("Cannot get node name.")
+		return cloudprovider.Zone{}, err
+	}
+
+	zone := cloudprovider.Zone{}
+	vsi, err := vs.getVSphereInstanceForServer(vs.cfg.Workspace.VCenterIP, ctx)
+
+	if err != nil {
+		glog.Errorf("Cannot connent to vsphere. Get zone for node %s error", nodeName)
+		return cloudprovider.Zone{}, err
+	}
+
+	dc, err := vclib.GetDatacenter(ctx, vsi.conn, vs.cfg.Workspace.Datacenter)
+	if err != nil {
+		glog.Errorf("Cannot connent to datacenter. Get zone for node %s error", nodeName)
+		return cloudprovider.Zone{}, err
+	}
+
+	vmHost, err := dc.GetHostByVMUUID(ctx, vs.vmUUID)
+	if err != nil {
+		glog.Errorf("Cannot find VM runtime host. Get zone for node %s error", nodeName)
+		return cloudprovider.Zone{}, err
+	}
+
+	client := vsi.conn
+	zoneManagers := tagmanagers.NewTagManagers(client)
+	err = zoneManagers.WithClient(ctx, func(client *tags.RestClient) error {
+
+		tags, err := client.ListAttachedTags(
+			ctx,
+			vmHost,
+		)
+
+		if err != nil {
+			glog.Errorf("Cannot list attached tags. Get zone for node %s error", nodeName)
+			return err
+		}
+
+		for _, value := range tags {
+			tag, err := client.GetTag(ctx, value)
+			if err != nil {
+				glog.Errorf("Get tag %s error", value)
+				return err
+			}
+			category, err := client.GetCategory(ctx, tag.CategoryID)
+			if err != nil {
+				glog.Errorf("Get category %s error", value)
+				return err
+			}
+
+			switch {
+
+			case category.Name == vs.cfg.Labels.Zone:
+				zone.FailureDomain = tag.Name
+
+			case category.Name == vs.cfg.Labels.Region:
+				zone.Region = tag.Name
+
+			default:
+				zone.FailureDomain = ""
+				zone.Region = ""
+			}
+		}
+
+		switch {
+		case zone.Region == "":
+			if vs.cfg.Labels.Zone != "" {
+				return fmt.Errorf("The zone in vSphere configuration file not match for node %s ", nodeName)
+			}
+			glog.Infof("No zones support for node %s error", nodeName)
+			return nil
+
+		case zone.FailureDomain == "":
+			if vs.cfg.Labels.Region != "" {
+				return fmt.Errorf("The zone in vSphere configuration file not match for node %s ", nodeName)
+			}
+			glog.Infof("No zones support for node %s error", nodeName)
+			return nil
+		}
+		return nil
+	})
+	if err != nil {
+		glog.Errorf("Get zone for node %s error", nodeName)
+		return cloudprovider.Zone{}, err
+	}
+	return zone, nil
+}
+
+func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.NodeName) (cloudprovider.Zone, error) {
+
+	return cloudprovider.Zone{}, nil
+
+}
+
+func (vs *VSphere) GetZoneByProviderID(ctx context.Context, providerID string) (cloudprovider.Zone, error) {
+
+	return cloudprovider.Zone{}, nil
+}
```

Comment:
`GetZoneByNodeName()` and `GetZoneByProviderID()` are called by out-tree cloud provider, which have not been tested yet. I have completed the code for those two funcs but for now I leave it just by returning an empty zone. If this is not a good way, please comment.

Reply:
Changed to `cloudprovider.NotImplemented` here.